### DISCUSSION TITLE
Page that won't let you continue until all questions are answered correctly

### DISCUSSION THREAD

**Discussion by slussinator** [ctreffe/alfred#247] [Alfred Q&A] [answered]
Dear community, 

I would like to implement a manipulation check in the form that participants have to answer all questions correctly efore continuing to the next page. Attached find a minimal experiment that should function as a basic framework to build on. I know how to "test" and count how many questions are answered correctly (see lower part of my code). I don't know however how to translate this into a way of letting participants continue or repeat the page. 

Thank you in advance! 
Sophie 

```python
import alfred3 as al
exp = al.Experiment()


ManipulationCheck_list_MC = [
    {"name": "MC_1_lo", "toplab": "This is true"},
    {"name": "MC_2_hi", "toplab": "This is false"}
    ]

@exp.member
class lo_ManipulationCheck(al.Page):  
    title = "Manipulation Check"

    def on_exp_access(self):

        for i in ManipulationCheck_list_MC:
            # The following two lines will skip the remaining iterations when the
            # item list runs out of items, giving the code way more flexibility.
            item = i
            self += al.SingleChoiceButtons(
                "True",
                "False",
                toplab=item['toplab'],
                name=item['name'],
                width="wide",
                button_width="140px",
                force_input=True,
                font_size="normal",
                position="center"
            )
    def on_first_hide(self):

        attention_score = 0  
        if self.exp.values.get(
                "MC_1_lo") == 1:  
            attention_score += 1
        if self.exp.values.get(
                "MC_2_hi") == 2:
            attention_score += 1
     


if __name__ == "__main__":
    exp.run()

```

**Comment by jobrachem** [comment]
Hi @slussinator , I am quite certain that this is possible. To go on, would you mind reducing your example to just two items and remove elements from the code that are not relevant to the question? I think this applies to the `al.Card` and `al.Value`. And if you define code blocks with the `python` tag, you get syntax highlighting, which also helps a lot. Like this:

    ```python
    import alfred3 as al
    ```

Leads to:

```python
import alfred3 as al
```

**Reply by slussinator** [reply]
Hi @jobrachem, thanks for your response. I have adjusted the code as requested. I hope everything is correct now!
Looking forward to your suggestions :)

**Reply by jobrachem** [reply]
Hi @slussinator, thank you! I do not want to be obnoxious, but your code example still contains more than 50% irrelevant lines. This is what a minimal example for this problem would look like:

```python
import alfred3 as al

exp = al.Experiment()

@exp.member
class Demo(al.Page):  

    def on_exp_access(self):
        self += al.SingleChoiceButtons("True", "False", toplab="true" ,name="item1")
        self += al.SingleChoiceButtons("True", "False", toplab="false", name="item2")

    def on_first_hide(self):

        attention_score = 0  
        if self.exp.values.get("item1") == 1:  
            attention_score += 1
        if self.exp.values.get("item2") == 2:
            attention_score += 1      
```

I stress this point, because this a "help me to help you" situation. If the code example is reduced to the absolute minimum, it is often very easy for me to see what could be a solution. On the other hand, if it is not really minimal, I have to read much more carefully. That takes more time for me and is more error prone. So you will always get faster and better answers if you use minimal examples. That being said, of course I see that you did already put a lot of effort into crafting the code for your example, and I thank you for that :blush:

I provide my answer separately below.

**Comment by jobrachem** [comment, marked as answer]
There are two ways to do this.

**Because Option 2 is not documented, I recommend Option 1.**

### Option 1: Page.custom_move()

You can use the [`Page.custom_move()`](https://alfredo3.psych.bio.uni-goettingen.de/docs/generated/alfred3.page.Page.custom_move.html#alfred3.page.Page.custom_move) hook. This hook is used to conduct manually specified movements. Usually, you would use `ExperimentSession.forward()` or `ExperimentSession.backward()` or `ExperimentSession.jump()` inside this method. But if it returns a "falsy" value like `False` or `None` *without* any custom movement, alfred will just stay on the same page. To fall back to alfreds normal movement system, you can define a case in which the method returns `True`. In your example, that could look like this:



```python
import alfred3 as al

exp = al.Experiment()

@exp.member
class Demo(al.Page):  

    def on_exp_access(self):

        self += al.SingleChoiceButtons("True", "False", toplab="true" ,name="item1")
        self += al.SingleChoiceButtons("True", "False", toplab="false", name="item2")

    def custom_move(self):

        attention_score = 0  
        if self.exp.values.get("item1") == 1:  
            attention_score += 1
        if self.exp.values.get("item2") == 2:
            attention_score += 1
        
        if attention_score < 2:
            self.exp.post_message("Test failed", level="danger")
            return False
        else:
            return True
     

if __name__ == "__main__":
    exp.run()
```

### Option 2: AbortMove

Alfred defines the exception `AbortMove`. If you raise this exception during the `Page.on_first_hide` hook, the movement will be aborted. This could look like this:

```python
import alfred3 as al
from alfred3.exceptions import AbortMove

exp = al.Experiment()

@exp.member
class Demo(al.Page):  

    def on_exp_access(self):

        self += al.SingleChoiceButtons("True", "False", toplab="true" ,name="item1")
        self += al.SingleChoiceButtons("True", "False", toplab="false", name="item2")

    def on_first_hide(self):

        attention_score = 0  
        if self.exp.values.get("item1") == 1:  
            attention_score += 1
        if self.exp.values.get("item2") == 2:
            attention_score += 1
        
        if attention_score < 2:
            self.exp.post_message("Test failed", level="danger")
            raise AbortMove
     


if __name__ == "__main__":
    exp.run()
```


**Because Option 2 is not documented, I recommend Option 1.**

**Reply by slussinator** [reply]
Thanks @jobrachem! This works perfectly. Exactly what I had in mind :) 
(I did go with Option 1 as you suggested).

**Reply by jobrachem** [reply]
@slussinator that's very nice to hear 😊
